src/plugins/intel_npu/tests/unit/npuw/permute.cpp

```diff
@@ -0,0 +1,255 @@
+// Copyright (C) 2024 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#ifdef HAVE_AVX2
+#    include "permute.hpp"
+
+namespace details {
+
+// Permute function.
+template <typename T>
+void permute(const T* input, T* output, const std::vector<size_t>& dims, const std::vector<size_t>& order) {
+    size_t ndim = dims.size();
+    assert(order.size() == ndim);
+
+    // Compute new dims.
+    std::vector<size_t> new_dims = details::reorder(dims, order);
+
+    // Compute strides.
+    std::vector<size_t> old_strides = details::compute_strides(dims);
+    std::vector<size_t> new_strides = details::compute_strides(new_dims);
+
+    size_t total = std::accumulate(dims.begin(), dims.end(), size_t(1), std::multiplies<size_t>());
+
+    // For each index in output, compute corresponding input index.
+    // Use a flat index and unravel it to multi-index in new order, then map to old order.
+    for (size_t out_idx = 0; out_idx < total; ++out_idx) {
+        // Unravel out_idx to multi-index in new_dims.
+        std::vector<size_t> new_multi(ndim);
+        size_t tmp = out_idx;
+        for (size_t i = 0; i < ndim; ++i) {
+            new_multi[i] = tmp / new_strides[i];
+            tmp = tmp % new_strides[i];
+        }
+        // Map new_multi to old_multi.
+        std::vector<size_t> old_multi(ndim);
+        for (size_t i = 0; i < ndim; ++i)
+            old_multi[order[i]] = new_multi[i];
+
+        // Compute flat input index.
+        size_t in_idx = 0;
+        for (size_t i = 0; i < ndim; ++i)
+            in_idx += old_multi[i] * old_strides[i];
+
+        if constexpr (std::is_same<T, int8_t>::value) {
```

Comment:
Good solution!

Let's just make it a bit more explicit:
```c++
template <typename T, bool Int4 = false>
void permute(const T* input, T* output, const std::vector<size_t>& dims, const std::vector<size_t>& order) {

...

static_assert(!Int4 || (Int4 && std::is_same_v<T, uint8_t>), "For int4 data input data must be of type uint8_t");
if constexpr (Int4) {
...
}
else
{
...
}
```

Just in case we would need to support `uint8_t` tensors which are not int4.